subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/TransformingAsyncArtifactListener.java

```diff
@@ -37,31 +39,38 @@
-        Map<ComponentArtifactIdentifier, TransformationOperation> artifactResults,
-        Map<File, TransformationOperation> fileResults,
-        ExecutionGraphDependenciesResolver dependenciesResolver
+        Map<ComponentArtifactIdentifier, TransformationResult> artifactResults,
+        Map<File, TransformationResult> fileResults,
+        ExecutionGraphDependenciesResolver dependenciesResolver,
+        TransformationNodeFactory transformationNodeFactory
     ) {
         this.artifactResults = artifactResults;
         this.actions = actions;
         this.transformation = transformation;
         this.delegate = delegate;
         this.fileResults = fileResults;
         this.dependenciesResolver = dependenciesResolver;
+        this.transformationNodeFactory = transformationNodeFactory;
     }
 
     @Override
     public void artifactAvailable(ResolvableArtifact artifact) {
         ComponentArtifactIdentifier artifactId = artifact.getId();
-        File file = artifact.getFile();
-        TransformationSubject initialSubject = TransformationSubject.initial(artifactId, file);
-        TransformationOperation operation = new TransformationOperation(transformation, initialSubject, dependenciesResolver);
-        artifactResults.put(artifactId, operation);
-        // We expect artifact transformations to be executed in a scheduled way,
-        // so at this point we take the result from the in-memory cache.
-        // Artifact transformations are always executed scheduled via the execution graph when the transformed component is declared as an input.
-        // Using the BuildOperationQueue here to only realize that the result of the transformation is from the in-memory has a performance impact,
-        // so we executing the (no-op) operation in place.
-        operation.run(null);
+        Try<TransformationSubject> resultIfCompleted = transformationNodeFactory.getResultIfCompleted(artifactId, transformation);
+        if (resultIfCompleted != null) {
+            artifactResults.put(artifactId, new PrecomputedTransformationResult(resultIfCompleted));
+        } else {
+            File file = artifact.getFile();
+            TransformationSubject initialSubject = TransformationSubject.initial(artifactId, file);
+            TransformationOperation operation = new TransformationOperation(transformation, initialSubject, dependenciesResolver);
+            artifactResults.put(artifactId, operation);
+            // We expect artifact transformations to be executed in a scheduled way,
+            // so at this point we take the result from the in-memory cache.
+            // Artifact transformations are always executed scheduled via the execution graph when the transformed component is declared as an input.
+            // Using the BuildOperationQueue here to only realize that the result of the transformation is from the in-memory has a performance impact,
+            // so we executing the (no-op) operation in place.
```

Comment:
I must have drunk something pretty strong when I wrote this as I haven't the faintest about what I wanted to actually say anymore. :)